src/main/java/org/motechproject/prevac/service/impl/VisitScheduleServiceImpl.java

```diff
@@ -173,41 +174,55 @@ private void setVisitPlannedDate(Visit baseVisit, Visit visit, VisitScheduleOffs
-        }
+            if (!visitDates.containsKey(SCREENING_DATE_KEY)) {
+                throw new VisitScheduleException("Couldn't calculate Planned Dates, because Participant didn't participate in screening visit");
+            }
 
-        LocalDate earliestDate;
-        LocalDate latestDate = screeningDate.plusDays(PrevacConstants.LATEST_DATE);
+            LocalDate earliestDate;
+            LocalDate latestDate = visitDates.get(SCREENING_DATE_KEY).plusDays(PrevacConstants.LATEST_DATE);
 
-        if (subject.getFemaleChildBearingAge() == null || !subject.getFemaleChildBearingAge()) {
-            earliestDate = screeningDate;
-        } else {
-            earliestDate = screeningDate.plusDays(PrevacConstants.EARLIEST_DATE_IF_FEMALE_CHILD_BEARING_AGE);
-        }
+            if (subject.getFemaleChildBearingAge() == null || !subject.getFemaleChildBearingAge()) {
+                earliestDate = visitDates.get(PRIME_VAC_DATE_KEY);
+            } else {
+                earliestDate = visitDates.get(PRIME_VAC_DATE_KEY).plusDays(PrevacConstants.EARLIEST_DATE_IF_FEMALE_CHILD_BEARING_AGE);
+            }
 
-        if (date.isBefore(earliestDate) || date.isAfter(latestDate)) {
-            throw new VisitScheduleException(String.format("The Prime Vaccination date should be between %s and %s but is %s",
-                    earliestDate, latestDate, date));
+            if (primeVacDate.isBefore(earliestDate) || primeVacDate.isAfter(latestDate)) {
+                throw new VisitScheduleException(String.format("The Prime Vaccination date should be between %s and %s but is %s",
+                    earliestDate, latestDate, primeVacDate));
+            }
         }
     }
 
-    private LocalDate getScreeningDate(Subject subject) {
-        LocalDate screeningDate = null;
+    private Map<String, LocalDate> getScreeningAndPrimeVacDates(Subject subject) {
+        Map<String, LocalDate> datesMap = new HashMap<>();
 
         for (Visit visit : subject.getVisits()) {
-            if (VisitType.SCREENING.equals(visit.getType())) {
-                screeningDate = visit.getDate();
+            if (VisitType.PRIME_VACCINATION_DAY.equals(visit.getType()) && visit.getDateProjected() != null) {
+                datesMap.put(PRIME_VAC_DATE_KEY, visit.getDateProjected());
+            }
+            if (VisitType.SCREENING.equals(visit.getType()) && visit.getDate() != null) {
+                datesMap.put(SCREENING_DATE_KEY, visit.getDate());
             }
         }
 
-        if (screeningDate == null) {
+        if (datesMap.isEmpty()) {
+            throw new VisitScheduleException(String.format("Couldn't save Planned Dates, because Participant with Id:" +
+                "%s didn't participate in screening visit and doesn't have Prime Vaccination visit planned.", subject.getSubjectId()));
+        }
+        if (!datesMap.containsKey(PRIME_VAC_DATE_KEY)) {
```

Comment:
I don't think it should be required, we can use screening date if it's null (like it was done before)